File: r/04.R
#!/usr/bin/env Rscript

library(digest)
library(magrittr)

mine <- function(input, test) {
  length <- nchar(test)
  getHash <- function(x) {
    input %>% paste0(x) %>%
      digest(algo = 'md5', serialize = FALSE) %>%
      substr(1, length)
  }

  i <- 1
  while (getHash(i) != test) {
    i <- i + 1
  }
  i
}

side.a <- function(input) {
  mine(input, '00000')
}

side.b <- function(input) {
  mine(input, '000000')
}

main <- function() {
  input <- 'ckczppom'
  cat(side.a(input), '\n')
  cat(side.b(input), '\n')
}

main()

File: js/04.js
#!/usr/bin/env node
'use strict'

const md5 = require('md5')
const fs = require('fs')

const mine = (input, test) => {
  const hash = (str) => md5(str).slice(0, test.length)

  let i = 0
  while (hash(input + i) !== test) ++i
  return i
}

const sideA = (str) => mine(str, '00000')

const sideB = (str) => mine(str, '000000')

const main = () => {
  const input = fs.readFileSync('../input/04.txt', 'utf8').trim()
  console.log(sideA(input))
  console.log(sideB(input))
  return
}

main()
